addons/mail/models/mail_channel.py

```diff
@@ -238,12 +376,53 @@ def message_get_recipient_values(self, notif_message=None, recipient_ids=None):
             }
         return super(Channel, self).message_get_recipient_values(notif_message=notif_message, recipient_ids=recipient_ids)
 
+    def _extract_values(self, message_type, **kwargs):
+        """ This method is used to compute moderation status before the creation of a message. 
+            For this operation the message's author email address is required.
+            This address is returned with status for other computations.
+        """
+        moderation_status = 'accepted'
+        email = ''
+        if self.moderation and message_type in ['email', 'comment']:
+            author_id = kwargs.get('author_id')
+            if author_id and isinstance(author_id, pycompat.integer_types):
+                author = self.env['res.partner'].browse([author_id])
+                email = author.email
+            elif author_id:
+                email = author_id.email
+            elif self.env.user:
+                email = self.env.user.email
+            else:
+                email_from = kwargs.get('email_from')
+                if email_from:
+                    email = tools.email_split(email_from)[0]
+            if self.env['mail.moderation.email'].search([('email', '=', email), ('status', '=', 'allow')]) or email in self.mapped('moderator_ids.email'):
+                moderation_status = 'accepted'
+            elif self.env['mail.moderation.email'].search([('email', '=', email), ('status', '=', 'ban')]):
```

Comment:
Instead of performing two search, you could do everything in one
```
status = self.env[].search([(channel_id, in, self.ids), (email, =, email)]).status or 'pending'
```
Then directly return status and use if in message_post (values will therefore be allow, ban or pending, meaning we use globally the same keys as used for the field itself, simplifying the code understanding).

Reply:
I think it is nice that way. The test are not exactly the same and it would require a translation function.